determined-ai/determined
# feat: Create Webhook Sender [WEB-213]

## Description
feat: Create Webhook Sender [WEB-213]

This PR also implements:
https://determinedai.atlassian.net/browse/WEB-217
https://determinedai.atlassian.net/browse/WEB-214
https://determinedai.atlassian.net/browse/WEB-212

This PR contains the backend updates needed to support webhooks in determined. A `webhook_event_que` table has been added which is the que for webhook events. When an experiment state change occurs, if any triggers exist for a webhook related to the state change then the appropriate payload for the webhook type is created and saved to the event que. An `event` shipper has been created that will pull events from the que and attempt to deliver events to the webhook URL. 

### Security and Signed Payload

Each Webhook request will include a signed payload that users can use to verify each event was sent from Determined. This will require that users define a `webhook_signing_key` for signing. If a webhook signing key is not provided then one will be generated for the user.

#### Setting the Key
The signing key can be set in the following ways:

1. In the master `yaml` configuration file by adding:
```
      # config_file is just a master.yaml
      config_file:
        webhook:
          signing_key: <signing_key_value>
```

2. By setting a `DET_WEBHOOK_SIGNING_KEY` environment variable. 
3. By specifying a `---security-webhook-signing-key` flag 

### Event Payloads

Currently we will support two separate types of Webhooks `Slack` and `Default`. A payload for a Default webhook will contain information about the event itself, the trigger for the event, and the appropriate entity that triggered the event. Since we only support experiment state changes for now all payloads will look like: 

```
{
  "event_id": "4cd26e62-60c6-4a8b-8d03-7629091a4ef5".   // unique event uuid,
  "event_type": "EXPERIMENT_STATE_CHANGE" // the trigger type for the event,
  "timestamp": 1665689991 // the timestamp of the event occurence,
  "condition": {. 
    "state": "COMPLETED". // the condition that triggered the event for now the condition will always be of this format and contain either "COMPLETED" or "ERRORED"
  },
  "event_data": {
    "experiment": {
      "id": 41, 
      "state": "COMPLETED",
      "name": "cifar10_pytorch_const",
      "duration": 41, // this is in seconds
      "resource_pool": "default",
      "slots": 0,
      "workspace": "test workspace", // name of the workspace
      "project": "test project" // name of the project
    }
  }
}
```

A `Base URL` configuration has been added that will be used to allow us to link users to Workspaces, Projects, and Experiments. It is discussed in the docs here: https://github.com/determined-ai/determined/blob/master/docs/integrations/notification/slack.rst



### Retrieving the Key

The key can be found in the cluster configuration. For example it will be returned in `api/v1/master/config`.

### Examples

Example Default Webhook Payload:
<img width="1233" alt="Screen Shot 2022-10-13 at 4 07 56 PM" src="https://user-images.githubusercontent.com/103522725/195710955-bb080009-a8fa-4dcd-8747-bceb0308df8c.png">

Example Slack Message:
![Screen Shot 2022-10-13 at 4 09 21 PM](https://user-images.githubusercontent.com/103522725/195711151-5b591cf9-2a5b-403e-bd65-637f596bf46f.png)

Example Test Default Payload:

<img width="1241" alt="Screen Shot 2022-10-13 at 4 21 10 PM" src="https://user-images.githubusercontent.com/103522725/195713069-2dcc804c-96ca-49be-8f6d-c199243baea1.png">

Example Test Slack Payload:
![Screen Shot 2022-10-13 at 4 52 22 PM](https://user-images.githubusercontent.com/103522725/195717871-dadfc1e4-9eaa-4ce8-9d97-3fc4cb432d63.png)

## Test Plan
1. Create a webhook of type Slack and Default in Determined. Add triggers for "Completed" and "Errored" experiments. For the Default webhook I suggest getting a URL from https://webhook.site. For a Slack URL you can use the Test Webhook Service slack application here: https://api.slack.com/apps/A040M5WLT3N/incoming-webhooks
2. Run an experiment that will either fail or complete. 
3. Check to ensure that both the default and slack destinations received a webhook request and are formatted correctly. 
4. Start an experiment that is in a project and set the cluster `WEBHOOK_BASE_URL`. Ensure that the Slack message correctly contains links to the experiment, workspace, and project. 
5. Check the cluster configuration and ensure that a `webhook_signing_key` has been set by default.
6. Specify a `webhook_signing_key`. Check the webhook configuration and ensure that the correct value is set. 
 



## Commentary (optional)




## Checklist
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.





[WEB-213]: https://determinedai.atlassian.net/browse/WEB-213?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
